### PR TITLE
Login

### LINKED ISSUE #82: Login (same for all users)
Add functionality for a user to login (same for admin, customer and technician).

This should probably be done in the "utility" package within the "service" package since it will be shared.

We'll pretend that encryption doesn't exist, so just check if the entered password is the same as the one in the database for the given email.

### PR DESCRIPTION
**Pull request by JDziewonski98**
Super basic login system to keep things moving. Is it sufficient for scope of this class? 
I imagine front-end having a selection like "I am a customer/admin/technician" on login page that determines UserType.

### CODE REVIEW

File: RepairShop-Backend/src/main/java/ca/mcgill/ecse321/repairshop/service/AuthenticationService.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    @Autowired
+    TechnicianRepository technicianRepository;
+
+    /**'
+     * Validates an email + password pair for a user type
+     * @param credentials DTO containing login credentials and user type
+     * @return true is login was successful otherwise false
+     */
+    @Transactional
+    public boolean logIn(LoginDto credentials) {
```
Comment (T-Syme) [inline RepairShop-Backend/src/main/java/ca/mcgill/ecse321/repairshop/service/AuthenticationService.java:33]: Does this method need to have "throws Exception" if the email cannot be found?

File: RepairShop-Backend/src/main/java/ca/mcgill/ecse321/repairshop/service/AuthenticationService.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+            case Technician:
+                return authenticateTechnician(credentials.getEmail(), credentials.getPassword());
+            case Admin:
+                return authenticateAdmin(credentials.getEmail(), credentials.getPassword());
+            case Customer:
+                return authenticateCustomer(credentials.getEmail(), credentials.getPassword());
+        }
+        return false;
+    }
+
```
Comment (T-Syme) [inline RepairShop-Backend/src/main/java/ca/mcgill/ecse321/repairshop/service/AuthenticationService.java:44]: Won't the authenticate[User] methods have a NullPointerException if the email isn't valid?
Maybe I'm missing something, but would tech = null, then (null).getPassword() would give you an exception?

File: RepairShop-Backend/src/test/java/ca/mcgill/ecse321/repairshop/service/TestAuthenticationService.java [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+    public void testInvalidAdminLogin() {
+        LoginDto loginDto = new LoginDto();
+        loginDto.setEmail(badUserEmail);
+        loginDto.setPassword(userPassword);
+        loginDto.setUserType(UserType.Admin);
+        try {
+            authenticationService.logIn(loginDto);
+            fail("Should throw an exception.");
+        } catch (Exception e) {
+            assertNull(e.getMessage());
```
Comment (T-Syme) [inline RepairShop-Backend/src/test/java/ca/mcgill/ecse321/repairshop/service/TestAuthenticationService.java]: Won't this always fail? Assuming there is an error message...

**Review by T-Syme — COMMENTED** [review]
I left a few comments about stuff I wasn't sure about

**Review by Shaswata79 — APPROVED** [review]

File: RepairShop-Backend/src/main/java/ca/mcgill/ecse321/repairshop/service/AuthenticationService.java (hunk 2 above)
Comment (JDziewonski98) [reply RepairShop-Backend/src/main/java/ca/mcgill/ecse321/repairshop/service/AuthenticationService.java:44]: it throws an exception when it isn't found by email, which is caught in the controller and returns httpstatus.NOTFOUND

File: RepairShop-Backend/src/test/java/ca/mcgill/ecse321/repairshop/service/TestAuthenticationService.java (hunk 3 above)
Comment (JDziewonski98) [reply RepairShop-Backend/src/test/java/ca/mcgill/ecse321/repairshop/service/TestAuthenticationService.java]: It throws an exception by design if the email is not found

Comment (JDziewonski98) [reply RepairShop-Backend/src/test/java/ca/mcgill/ecse321/repairshop/service/TestAuthenticationService.java]: I wont PR tests that always fail haha

Comment (T-Syme) [reply RepairShop-Backend/src/test/java/ca/mcgill/ecse321/repairshop/service/TestAuthenticationService.java]: Right, but if it throws an exception then wouldn't e.getMessage() return a message that is not null?

**Review by sandyl289 — APPROVED** [review]

File: RepairShop-Backend/src/test/java/ca/mcgill/ecse321/repairshop/service/TestAuthenticationService.java (hunk 3 above)
Comment (JDziewonski98) [reply RepairShop-Backend/src/test/java/ca/mcgill/ecse321/repairshop/service/TestAuthenticationService.java]: Yeah but I'm just mocking the exception in the tests. Ill change it to mock the actual type of error it throws, whatever it is, and assert on that

**Pull request merged**